src/util/util.ts

```diff
@@ -37,3 +37,30 @@ export const isSubstringOf = (substring: string, source: string): boolean =>
   pipe(source, S.toLowerCase, S.includes(S.toLowerCase(substring)))
 
 export const isError = (value: unknown): value is Error => value instanceof Error
+
+export const escapeSparqlStr = (str?: string): string => {
+  const escapeCharactersMap: Record<string, string> = {
+    '\t': '\\t',
+    '\n': '\\n',
+    '\r': '\\r',
+    '\b': '\\b',
+    '\f': '\\f',
+    '"': '\\"',
+    "'": "\\'",
+    '\\': '\\\\'
+  }
+
+  return pipe(
+    str,
+    O.fromNullable,
+    O.map(s =>
+      pipe(
+        s,
+        Array.from,
```

Comment:
Use `A.of` instead. This will help resolve the type inference you had.

Reply:
I tried to apply all your recommendations, and it worked pretty well except for this one, I ran into some typing issues. The `A.of` function takes a value and puts it into an array, whereas `Array.from` converts an iterable (or array-like) object into an array. Since I needed to convert a string into an array of its characters, I didn't find a cleaner solution than `Array.from`.

Is there a way to use `A.of` in this case? Thanks

Reply:
Ah, my bad. I see your point. `A.of just` creates an array with the string as unique element. It could be done in pure fp-ts, but it adds some boilerplate code:

```ts
pipe(
  s,
  S.split(''),
  A.of
)
```

If you want to continue down this path, you'll need to add a utility function that encapsulates this code. But I'll let you decide.

Reply:
Using `S.split('')` seemed more natural to me, but we need to deal with read-only Array.
And my bad, I forgot about using directly `ReadonlyArray` with `map`.
Latest commit was updated by and `Array.from` was replaced with `S.split('')`.
```ts
S.split(''),
RA.map(c =>
```

Reply:
You can use [`S.split('')`](https://gcanti.github.io/fp-ts/modules/string.ts.html#split) which returns an `ReadonlyNonEmptyArray`  of characters. But it's not that impressive.

```ts
  pipe(
    s,
    S.split(''),
  )
```

if I combine everything together, we should get something like this:

```ts
export const escapeSparqlStr = (str?: string): string => {
  const escapeCharactersMap: Record<string, string> = {
    '\t': '\\t',
    '\n': '\\n',
    '\r': '\\r',
    '\b': '\\b',
    '\f': '\\f',
    '"': '\\"',
    "'": "\\'",
    '\\': '\\\\'
  }

  const lookupEscaperCharacter = (c: string): string =>
    pipe(
      escapeCharactersMap,
      R.lookup(c),
      O.getOrElse(constant(c))
    )

  return pipe(
    str,
    O.fromNullable,
    O.getOrElse(constant(S.empty)),
    S.split(S.empty),
    RA.map(lookupEscaperCharacter),
    RA.reduce(S.Monoid.empty, S.Monoid.concat)
  )
}
```

There must be room for improvement, but I'm running out of time. I leave this in your hands for integration, at your convenience.

Reply:
That's a fantastic way to make it easier to read and understand. We can use your solution. Thanks!